HTTP/2 Client infinite reconnection loop

**Jetty version(s)**

Isolated to 12.0.23 and newer. (12.0.22 is unaffected)

**Jetty Environment**
N/A

**HTTP version**
HTTP/2

**Java version/vendor** `(use: java -version)`
JDK 21

**OS type/version**
Linux

**Description**

When using a Jetty WebSocket client with HttpClientTransportDynamic and HTTP/2 enabled, if TLS negotiation fails the client goes into an infinite reconnection loop.

**How to reproduce?**

The below code replicates the bug, in Jetty 12.0.23 and newer it gets stuck:

```java
package org.example;

import com.sun.net.httpserver.HttpsConfigurator;
import com.sun.net.httpserver.HttpsParameters;
import com.sun.net.httpserver.HttpsServer;
import lombok.extern.slf4j.Slf4j;
import org.bouncycastle.asn1.x500.X500Name;
import org.bouncycastle.cert.X509CertificateHolder;
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
import org.bouncycastle.cert.jcajce.JcaX509v3CertificateBuilder;
import org.bouncycastle.operator.ContentSigner;
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.transport.HttpClientConnectionFactory;
import org.eclipse.jetty.client.transport.HttpClientTransportDynamic;
import org.eclipse.jetty.http2.client.HTTP2Client;
import org.eclipse.jetty.http2.client.transport.ClientConnectionFactoryOverHTTP2;
import org.eclipse.jetty.io.ClientConnectionFactory;
import org.eclipse.jetty.io.ClientConnector;
import org.eclipse.jetty.websocket.api.Session;
import org.eclipse.jetty.websocket.client.WebSocketClient;

import java.math.BigInteger;
import java.net.InetSocketAddress;
import java.net.URI;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.KeyStore;
import java.security.SecureRandom;
import java.security.cert.Certificate;
import java.security.cert.X509Certificate;
import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Date;
import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.TrustManagerFactory;

@Slf4j
public class TlsDeadlock {

    public static class ClientMain {
        // Run this with -Djdk.tls.namedGroups="secp256r1,secp384r1,secp521r1" to restrict named groups
        public static void main(String[] args) throws Exception {
            assert System.getProperty("jdk.tls.namedGroups").equals("secp256r1,secp384r1,secp521r1");
            ClientConnector clientConnector = new ClientConnector();
            ClientConnectionFactory.Info http1 = HttpClientConnectionFactory.HTTP11;
            ClientConnectionFactory.Info http2 = new ClientConnectionFactoryOverHTTP2.HTTP2(new HTTP2Client(clientConnector));
            HttpClient httpClient = new HttpClient(new HttpClientTransportDynamic(clientConnector, http1, http2));
            WebSocketClient client = new WebSocketClient(httpClient);
            client.setConnectTimeout(1000);
            client.setIdleTimeout(Duration.ofSeconds(1));
            client.start();
            System.out.println("Attempting to connect");
            // This should fail due to TLS negotiation
            client.connect(new Listener(), URI.create("wss://localhost:8999")).get();
            throw new RuntimeException("Unreachable");
        }
    }

    public static class Listener implements Session.Listener { }

    public static class ServerMain {
        public static void main(String[] args) throws Exception {
            // Create a self signed certificate
            KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
            keyGen.initialize(2048);
            KeyPair keyPair = keyGen.generateKeyPair();
            X500Name issuer = new X500Name("CN=WebSocketTest");
            BigInteger serial = BigInteger.valueOf(System.currentTimeMillis());
            Date notBefore = Date.from(Instant.now().minus(1, ChronoUnit.DAYS));
            Date notAfter = Date.from(Instant.now().plus(3650, ChronoUnit.DAYS));
            JcaX509v3CertificateBuilder certBuilder = new JcaX509v3CertificateBuilder(issuer, serial, notBefore, notAfter, issuer, keyPair.getPublic());
            ContentSigner signer = new JcaContentSignerBuilder("SHA256withRSA").build(keyPair.getPrivate());
            X509CertificateHolder certHolder = certBuilder.build(signer);
            X509Certificate cert = new JcaX509CertificateConverter().getCertificate(certHolder);
            KeyStore keyStore = KeyStore.getInstance("JKS");
            keyStore.load(null, null);
            keyStore.setKeyEntry("alias", keyPair.getPrivate(), "changeit".toCharArray(), new Certificate[]{cert});

            KeyManagerFactory kmf = KeyManagerFactory.getInstance("SunX509");
            kmf.init(keyStore, "changeit".toCharArray());
            TrustManagerFactory tmf = TrustManagerFactory.getInstance("SunX509");
            tmf.init(keyStore);

            // Create HTTPS Server
            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(kmf.getKeyManagers(), tmf.getTrustManagers(), new SecureRandom());
            HttpsServer server = HttpsServer.create();
            server.setHttpsConfigurator(new HttpsConfigurator(sslContext) {
                @Override
                public void configure(HttpsParameters params) {
                    SSLParameters sslParameters = sslContext.getDefaultSSLParameters();
                    sslParameters.setNamedGroups(new String[] {"x25519"});
                    sslParameters.setCipherSuites(new String[] {"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"});
                    params.setSSLParameters(sslParameters);
                }
            });
            server.bind(new InetSocketAddress("0.0.0.0", 8999), 100);
            server.start();
            System.out.println("Listening on " + server.getAddress());
        }
    }

}
```

Whereas in Jetty 12.0.22 it throws a TLS exception as expected:

```
Attempting to connect
Exception in thread "main" java.util.concurrent.ExecutionException: javax.net.ssl.SSLHandshakeException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.example.TlsDeadlock$ClientMain.main(TlsDeadlock.java:58)
Caused by: javax.net.ssl.SSLHandshakeException
	at org.eclipse.jetty.io.ssl.SslConnection$SslEndPoint.handshakeFailed(SslConnection.java:982)
	at org.eclipse.jetty.io.ssl.SslConnection$SslEndPoint.flush(SslConnection.java:1210)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:275)
```

It also works as expected if removing HTTP/2 from the connector.

Hi @lorban 
Thanks for working on this.
Yes I can confirm after compiling that branch I see this exception which is an improvement:

```
Exception in thread "main" java.util.concurrent.ExecutionException: org.eclipse.jetty.io.EofException
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.eclipse.jetty.client.transport.HttpRequest.send(HttpRequest.java:731)
	at org.eclipse.jetty.client.HttpClient.GET(HttpClient.java:379)
	at org.eclipse.jetty.client.HttpClient.GET(HttpClient.java:364)
	at org.example.TlsDeadlock$ClientMain.main(TlsDeadlock.java:55)
Caused by: org.eclipse.jetty.io.EofException
	at org.eclipse.jetty.io.NegotiatingClientConnection.onFillable(NegotiatingClientConnection.java:102)
	at org.eclipse.jetty.io.AbstractConnection$FillableCallback.succeeded(AbstractConnection.java:319)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ssl.SslConnection$SslEndPoint.onFillable(SslConnection.java:614)
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:415)
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:156)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:54)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:492)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.pcRunTask(AdaptiveExecutionStrategy.java:414)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:398)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:255)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:196)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:1009)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1239)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1194)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

I hope the final fix should show the real exception that we get when HTTP2 is disabled?

```
Exception in thread "main" java.util.concurrent.ExecutionException: javax.net.ssl.SSLHandshakeException: Abruptly closed by peer
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.eclipse.jetty.client.transport.HttpRequest.send(HttpRequest.java:731)
	at org.eclipse.jetty.client.HttpClient.GET(HttpClient.java:379)
	at org.eclipse.jetty.client.HttpClient.GET(HttpClient.java:364)
	at org.example.TlsDeadlock$ClientMain.main(TlsDeadlock.java:55)
Caused by: javax.net.ssl.SSLHandshakeException: Abruptly closed by peer
```

--- a/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
+++ b/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
@@ -17,8 +17,6 @@ import java.net.ServerSocket;
 import java.net.Socket;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicInteger;
-import java.util.concurrent.atomic.AtomicReference;
 import javax.net.ssl.SSLHandshakeException;
 
 import org.eclipse.jetty.client.HttpClient;
@@ -26,30 +24,32 @@ import org.eclipse.jetty.client.transport.HttpClientConnectionFactory;
 import org.eclipse.jetty.client.transport.HttpClientTransportDynamic;
 import org.eclipse.jetty.http2.client.HTTP2Client;
 import org.eclipse.jetty.http2.client.transport.ClientConnectionFactoryOverHTTP2;
+import org.eclipse.jetty.http2.client.transport.HttpClientTransportOverHTTP2;
 import org.eclipse.jetty.io.ClientConnectionFactory;
 import org.eclipse.jetty.io.ClientConnector;
 import org.eclipse.jetty.util.component.LifeCycle;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
 
 import static org.awaitility.Awaitility.await;
-import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.jupiter.api.Assertions.assertInstanceOf;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
 public class BrokenServerNegotiationTest
 {
-    @Test
-    public void testAlpnNegociationWithBrokenServer() throws Exception
+    private volatile ServerSocket serverSocket;
+
+    public int startServer() throws Exception
     {
-        AtomicReference<ServerSocket> serverSocketRef = new AtomicReference<>();
-        AtomicInteger port = new AtomicInteger();
+        if (serverSocket != null)
+            return serverSocket.getLocalPort();
         Thread server = new Thread(() ->
         {
             try
             {
-                ServerSocket serverSocket = new ServerSocket(0);
-                serverSocketRef.set(serverSocket);
-                port.set(serverSocket.getLocalPort());
+                serverSocket = new ServerSocket(0);
                 while (true)
                 {
                     Socket accept = serverSocket.accept();
@@ -62,20 +62,51 @@ public class BrokenServerNegotiationTest
             }
         });
         server.start();
-        await().atMost(5, TimeUnit.SECONDS).until(port::get, greaterThan(0));
+        await().atMost(5, TimeUnit.SECONDS).until(() -> serverSocket, notNullValue());
+        return serverSocket.getLocalPort();
+    }
 
+    @AfterEach
+    public void stopServer() throws Exception
+    {
+        if (serverSocket != null)
+            serverSocket.close();
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    public void testAlpnNegotiation(boolean useAlpn) throws Exception
+    {
+        int port = startServer();
         ClientConnector clientConnector = new ClientConnector();
         HTTP2Client h2Client = new HTTP2Client();
+        h2Client.setUseALPN(useAlpn);
         ClientConnectionFactoryOverHTTP2.HTTP2 http2 = new ClientConnectionFactoryOverHTTP2.HTTP2(h2Client);
         ClientConnectionFactory.Info http1 = HttpClientConnectionFactory.HTTP11;
         HttpClientTransportDynamic transportDynamic = new HttpClientTransportDynamic(clientConnector, http1,  http2);
         HttpClient client = new HttpClient(transportDynamic);
         client.start();
 
-        ExecutionException ee = assertThrows(ExecutionException.class, () -> client.GET("https://localhost:" + port.get() + "/hello"));
+        ExecutionException ee = assertThrows(ExecutionException.class, () -> client.GET("https://localhost:" + port + "/hello"));
         assertInstanceOf(SSLHandshakeException.class, ee.getCause());
 
         LifeCycle.stop(client);
-        serverSocketRef.get().close();
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    public void testH2Only(boolean useAlpn) throws Exception
+    {
+        int port = startServer();
+        HTTP2Client h2Client = new HTTP2Client();
+        h2Client.setUseALPN(useAlpn);
+        HttpClientTransportOverHTTP2 httpClientTransportOverHTTP2 = new HttpClientTransportOverHTTP2(h2Client);
+        HttpClient client = new HttpClient(httpClientTransportOverHTTP2);
+        client.start();
+
+        ExecutionException ee = assertThrows(ExecutionException.class, () -> client.GET("https://localhost:" + port + "/hello"));
+        assertInstanceOf(SSLHandshakeException.class, ee.getCause());
+
+        LifeCycle.stop(client);
     }
 }

--- a/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
+++ b/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
@@ -13,6 +13,8 @@
 
 package org.eclipse.jetty.http2.tests;
 
+import javax.net.ssl.SSLHandshakeException;
+
 import java.net.ServerSocket;
 import java.net.Socket;
 import java.util.concurrent.ExecutionException;
@@ -73,7 +75,7 @@ public class BrokenServerNegotiationTest
         client.start();
 
         ExecutionException ee = assertThrows(ExecutionException.class, () -> client.GET("https://localhost:" + port.get() + "/hello"));
-        assertInstanceOf(EofException.class, ee.getCause());
+        assertInstanceOf(SSLHandshakeException.class, ee.getCause());
 
         LifeCycle.stop(client);
         serverSocketRef.get().close();

--- a/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
+++ b/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
@@ -19,6 +19,7 @@ import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
+
 import org.eclipse.jetty.client.HttpClient;
 import org.eclipse.jetty.client.transport.HttpClientConnectionFactory;
 import org.eclipse.jetty.client.transport.HttpClientTransportDynamic;

--- a/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
+++ b/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
@@ -1,0 +1,80 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.http2.tests;
+
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.transport.HttpClientConnectionFactory;
+import org.eclipse.jetty.client.transport.HttpClientTransportDynamic;
+import org.eclipse.jetty.http2.client.HTTP2Client;
+import org.eclipse.jetty.http2.client.transport.ClientConnectionFactoryOverHTTP2;
+import org.eclipse.jetty.io.ClientConnectionFactory;
+import org.eclipse.jetty.io.ClientConnector;
+import org.eclipse.jetty.io.EofException;
+import org.eclipse.jetty.util.component.LifeCycle;
+import org.junit.jupiter.api.Test;
+
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class BrokenServerNegotiationTest
+{
+    @Test
+    public void testAlpnNegociationWithBrokenServer() throws Exception
+    {
+        AtomicReference<ServerSocket> serverSocketRef = new AtomicReference<>();
+        AtomicInteger port = new AtomicInteger();
+        Thread server = new Thread(() ->
+        {
+            try
+            {
+                ServerSocket serverSocket = new ServerSocket(0);
+                serverSocketRef.set(serverSocket);
+                port.set(serverSocket.getLocalPort());
+                while (true)
+                {
+                    Socket accept = serverSocket.accept();
+                    accept.close();
+                }
+            }
+            catch (Exception e)
+            {
+                // ignore, time to shut down
+            }
+        });
+        server.start();
+        await().atMost(5, TimeUnit.SECONDS).until(port::get, greaterThan(0));
+
+        ClientConnector clientConnector = new ClientConnector();
+        HTTP2Client h2Client = new HTTP2Client();
+        ClientConnectionFactoryOverHTTP2.HTTP2 http2 = new ClientConnectionFactoryOverHTTP2.HTTP2(h2Client);
+        ClientConnectionFactory.Info http1 = HttpClientConnectionFactory.HTTP11;
+        HttpClientTransportDynamic transportDynamic = new HttpClientTransportDynamic(clientConnector, http1,  http2);
+        HttpClient client = new HttpClient(transportDynamic);
+        client.start();
+
+        ExecutionException ee = assertThrows(ExecutionException.class, () -> client.GET("https://localhost:" + port.get() + "/hello"));
+        assertInstanceOf(EofException.class, ee.getCause());
+
+        LifeCycle.stop(client);
+        serverSocketRef.get().close();
+    }
+}

--- a/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
+++ b/jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/BrokenServerNegotiationTest.java
@@ -13,14 +13,13 @@
 
 package org.eclipse.jetty.http2.tests;
 
-import javax.net.ssl.SSLHandshakeException;
-
 import java.net.ServerSocket;
 import java.net.Socket;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
+import javax.net.ssl.SSLHandshakeException;
 
 import org.eclipse.jetty.client.HttpClient;
 import org.eclipse.jetty.client.transport.HttpClientConnectionFactory;
@@ -29,7 +28,6 @@ import org.eclipse.jetty.http2.client.HTTP2Client;
 import org.eclipse.jetty.http2.client.transport.ClientConnectionFactoryOverHTTP2;
 import org.eclipse.jetty.io.ClientConnectionFactory;
 import org.eclipse.jetty.io.ClientConnector;
-import org.eclipse.jetty.io.EofException;
 import org.eclipse.jetty.util.component.LifeCycle;
 import org.junit.jupiter.api.Test;
 

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
@@ -99,8 +99,7 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
                 }
                 else if (filled < 0)
                 {
-                    failure = new SSLHandshakeException("Abruptly closed by peer");
-                    break;
+                    throw new SSLHandshakeException("Abruptly closed by peer");
                 }
                 else if (filled == 0)
                 {

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
@@ -13,6 +13,7 @@
 
 package org.eclipse.jetty.io;
 
+import java.io.IOException;
 import java.util.Map;
 import java.util.concurrent.Executor;
 import java.util.concurrent.TimeoutException;
@@ -20,7 +21,6 @@ import javax.net.ssl.SSLEngine;
 import javax.net.ssl.SSLHandshakeException;
 
 import org.eclipse.jetty.util.BufferUtil;
-import org.eclipse.jetty.util.ExceptionUtil;
 import org.eclipse.jetty.util.Promise;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -61,7 +61,7 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
     protected void completed(String protocol)
     {
         this.protocol = protocol;
-        completed = true;
+        this.completed = true;
     }
 
     @Override
@@ -79,8 +79,7 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
         catch (Throwable x)
         {
             close();
-            // TODO: should we not fail the promise in the context here?
-            throw ExceptionUtil.asRuntimeException(x);
+            failConnectionPromise(x);
         }
     }
 
@@ -98,7 +97,7 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
                     replaceConnection();
                     break;
                 }
-                if (filled < 0)
+                else if (filled < 0)
                 {
                     failure = new SSLHandshakeException("Abruptly closed by peer");
                     break;
@@ -119,23 +118,7 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
         {
             LOG.atDebug().setCause(failure).log("Unable to fill from endpoint");
             close();
-            @SuppressWarnings("unchecked")
-            Promise<Connection> promise = (Promise<Connection>)context.get(ClientConnector.CONNECTION_PROMISE_CONTEXT_KEY);
-            promise.failed(failure);
-        }
-    }
-
-    private void replaceConnection()
-    {
-        EndPoint endPoint = getEndPoint();
-        try
-        {
-            endPoint.upgrade(connectionFactory.newConnection(endPoint, context));
-        }
-        catch (Throwable x)
-        {
-            LOG.atDebug().setCause(x).log("Unable to replace connection");
-            close();
+            failConnectionPromise(failure);
         }
     }
 
@@ -143,11 +126,7 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
     public boolean onIdleExpired(TimeoutException timeout)
     {
         getEndPoint().close(timeout);
-
-        @SuppressWarnings("unchecked")
-        Promise<Connection> promise = (Promise<Connection>)context.get(ClientConnector.CONNECTION_PROMISE_CONTEXT_KEY);
-        promise.failed(timeout);
-
+        failConnectionPromise(timeout);
         return false;
     }
 
@@ -157,5 +136,17 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
         // Gentler close for SSL.
         getEndPoint().shutdownOutput();
         super.close();
+    }
+
+    private void replaceConnection() throws IOException
+    {
+        getEndPoint().upgrade(connectionFactory.newConnection(getEndPoint(), context));
+    }
+
+    private void failConnectionPromise(Throwable failure)
+    {
+        @SuppressWarnings("unchecked")
+        Promise<Connection> promise = (Promise<Connection>)context.get(ClientConnector.CONNECTION_PROMISE_CONTEXT_KEY);
+        promise.failed(failure);
     }
 }

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
@@ -140,7 +140,8 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
 
     private void replaceConnection() throws IOException
     {
-        getEndPoint().upgrade(connectionFactory.newConnection(getEndPoint(), context));
+        EndPoint endPoint = getEndPoint();
+        endPoint.upgrade(connectionFactory.newConnection(endPoint, context));
     }
 
     private void failConnectionPromise(Throwable failure)

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
@@ -13,11 +13,11 @@
 
 package org.eclipse.jetty.io;
 
-import java.io.IOException;
 import java.util.Map;
 import java.util.concurrent.Executor;
 import java.util.concurrent.TimeoutException;
 import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLHandshakeException;
 
 import org.eclipse.jetty.util.BufferUtil;
 import org.eclipse.jetty.util.ExceptionUtil;
@@ -87,40 +87,41 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
     @Override
     public void onFillable()
     {
-        while (true)
-        {
-            int filled = fill();
-            if (completed)
-            {
-                replaceConnection();
-                break;
-            }
-            if (filled < 0)
-            {
-                @SuppressWarnings("unchecked")
-                Promise<Connection> promise = (Promise<Connection>)context.get(ClientConnector.CONNECTION_PROMISE_CONTEXT_KEY);
-                promise.failed(new EofException());
-                break;
-            }
-            else if (filled == 0)
-            {
-                fillInterested();
-                break;
-            }
-        }
-    }
-
-    private int fill()
-    {
+        Throwable failure = null;
         try
         {
-            return getEndPoint().fill(BufferUtil.EMPTY_BUFFER);
+            while (true)
+            {
+                int filled = getEndPoint().fill(BufferUtil.EMPTY_BUFFER);
+                if (completed)
+                {
+                    replaceConnection();
+                    break;
+                }
+                if (filled < 0)
+                {
+                    failure = new SSLHandshakeException("Abruptly closed by peer");
+                    break;
+                }
+                else if (filled == 0)
+                {
+                    fillInterested();
+                    break;
+                }
+            }
         }
-        catch (IOException x)
+        catch (Throwable x)
         {
-            LOG.atDebug().setCause(x).log("Unable to fill from endpoint");
+            failure = x;
+        }
+
+        if (failure != null)
+        {
+            LOG.atDebug().setCause(failure).log("Unable to fill from endpoint");
             close();
-            return -1;
+            @SuppressWarnings("unchecked")
+            Promise<Connection> promise = (Promise<Connection>)context.get(ClientConnector.CONNECTION_PROMISE_CONTEXT_KEY);
+            promise.failed(failure);
         }
     }
 

--- a/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
+++ b/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/NegotiatingClientConnection.java
@@ -90,12 +90,19 @@ public abstract class NegotiatingClientConnection extends AbstractConnection.Non
         while (true)
         {
             int filled = fill();
-            if (completed || filled < 0)
+            if (completed)
             {
                 replaceConnection();
                 break;
             }
-            if (filled == 0)
+            if (filled < 0)
+            {
+                @SuppressWarnings("unchecked")
+                Promise<Connection> promise = (Promise<Connection>)context.get(ClientConnector.CONNECTION_PROMISE_CONTEXT_KEY);
+                promise.failed(new EofException());
+                break;
+            }
+            else if (filled == 0)
             {
                 fillInterested();
                 break;

